Http client add url support

ohamuy opened the pull request:
HTTPClient : Support for full URL argument

Fixes # [812](https://github.com/open-telemetry/opentelemetry-cpp/issues/812)

Overloaded the CreateSession function allowing for url input:
Added it in the header file http_client.h
Implementation in http_client_curl.h
The function allows for url input instead of just host & port, utilizing the url parser in common folder. It returns an empty/non-existent host, and default port(80) if the parsing fails, and continues similarly to the original function otherwise.
One Unit test was changed to test the overloading function, but the overall functionality is not changed as the other original function parameters are tested in other unit tests. All tests passed, file builds successfully.
Lastly, the default port of 80 was removed from the original function because if the user ends up not providing a port, it will go through the other overloaded function and the url parser gives it a default port of 80

@lalitb @reyang

ThomsonTan commented:
@ohamuy thanks for your contribution. Could you please fix the format?

ohamuy commented:
@ThomsonTan Done! Also should I add to the change log for this commit?

maxgolov commented on ext/include/opentelemetry/ext/http/client/http_client.h:
```diff
@@ -227,8 +227,11 @@ class HttpClient
   virtual std::shared_ptr<Session> CreateSession(nostd::string_view host,
-                                                 uint16_t port = 80) noexcept = 0;
-  virtual bool CancelAllSessions() noexcept                                   = 0;
+                                                 uint16_t port) noexcept = 0;
```
Do we still need the port parameter here, if the URL itself can contain scheme and port?
Should we rename the `host` parameter to `url` or `uri` ?

maxgolov commented on ext/include/opentelemetry/ext/http/client/http_client.h:
```diff
@@ -227,8 +227,11 @@ class HttpClient
-  virtual bool CancelAllSessions() noexcept                                   = 0;
+                                                 uint16_t port) noexcept = 0;
+
+  virtual std::shared_ptr<Session> CreateSession(nostd::string_view url) noexcept = 0;
```
Okay, I see what you are doing. But I am not sure - maybe we should just deprecate the first method?

ohamuy replied:
Yes exactly, I just overloaded the function because that is what the @lalitb (the original issue creator) said might have been good. If its alright I can just simply remove the old parameter and just leave the new one since its also consistent with the `GET` and `POST` commands. But that is definitely a decision for the maintainers so when they give their input we can see whats the best course of action. @lalitb  @reyang

lalitb replied:
Yes, it should be fine removing the original existing method as @maxgolov suggested.

lalitb reviewed (approved):
LGTM once the proposed changes are done. Thanks for the contribution.

ohamuy commented:
@lalitb @maxgolov So I implemented the changes. Final change list:

-Deleted old `CreateSession` implementation
-Changed HTTPClient tests to match the correct parameters of the new function (all tests passed)
-Original constructor for `Session` was greatly simplified to one line as now we can pass in the scheme of `parsedUrl` 
-Added scheme parameter for `Session` constructor

ohamuy commented:
Ok so apparently removing the original function broke a lot of tests, what should I do?

maxgolov commented on .vscode/settings.json:
```diff
@@ -0,0 +1,89 @@
+{
```
Do we need this?

ohamuy replied:
I removed this in the most recent commit! it was an accidental file commited

maxgolov reviewed (approved):
LGTM, but I think your Visual Studio code settings is probably something that we'd rather not add in here?

maxgolov commented:
> Ok so apparently removing the original function broke a lot of tests, what should I do?

Doesn't look like _a lot_ - 3 files with tests... Fix them? 😃 

![image](https://user-images.githubusercontent.com/34072974/121487234-bdad8b80-c986-11eb-8112-403440bc8137.png)

lalitb commented:
Thank you for your contribution. Looks good to merge now :)
